### DISCUSSION TITLE
POCL - Pimcore Open Core License

### DISCUSSION THREAD

**Discussion by dpfaffenbauer** [pimcore/pimcore#17567] [General]
I have some questions about the POCL License that is used for Studio UI and hope someone from Pimcore can answer them :)

### SaaS/Hosting:
The POCL states that:

> 2.2 The Customer may not offer the Open Core Software as a hosted or managed service by granting
third parties access to a significant part of the features or functions of the Open Core Software.

Basically means SaaS right? If I host it for my customers, that is fine. But if I offer managed hosting for other Pimcore Agencies which they use it for their customers (= 3rdparty), its not allowed?

### POCL is GPL incompatible. 

But how do we then license our Extensions that use Studio? Do we have to License different pieces of the code differentially or make different Modules with different licenses? 🤯 

Like you do in Studio? Everything under `assets` and `public/build` is POCL the Rest is GPLv3? Like that the react code is POCL and the rest is GPLv3? What if someone includes 3rdparty libs that are GPLv3 in react? is that even possible?

Licenses are a very weird and confusing topic. 

### GPLv3, POCL and PCL 
Can every Extension License to these 3 Licenses as they wish? I mean, can I license my Extensions to PCL and POCL without acknowledgement from Pimcore?

**Comment by kingjia90** [comment]
@drietsch loves this topic, i am sure he is already on the way to reply to you 😄

**Reply by dpfaffenbauer** [reply]
I know he does @drietsch :heart:

**Comment by drietsch** [comment]
Hello Dominik!

Thanks for your questions. Let me quickly clarify.
### SaaS/Hosting:

> 2.2 The Customer may not offer the Open Core Software as a hosted or managed service by granting
> third parties access to a significant part of the features or functions of the Open Core Software.

Yes, you're right. The clause in the POCL restricts offering Pimcore as a Software as a Service (SaaS), Platform as a Service (PaaS) or managed service to third parties under certain conditions.

To clarify:

If you host Pimcore (including any POCL components) for your direct customers and provide them access to its features for their own usage, that is allowed under the license. However, hosters are not allowed to resell Pimcore (using its name or branding) or offer managed hosting or services to other Pimcore agencies or resellers, who would then provide it to their customers. This would be considered offering Pimcore as a SaaS, which is not permitted under the POCL.

So, in your example:

- Hosting it for your direct customers = Allowed.
- Offering hosting or a managed service to other agencies or resellers for them to sell to their clients = Not Allowed.

The key restriction is aimed at preventing unauthorized resale or repackaging of Pimcore, while still allowing businesses to use their preferred hosters for their own purposes.

### POCL is GPL incompatible.

Licensing, especially in open-source projects with multiple components, can indeed be tricky to navigate. Let's simplify the key points:

**POCL Compatibility:**
Anything under the Pimcore Open Core License (POCL) must remain under POCL and cannot use a copyleft license like GPLv3. Open-source licenses like MIT, Apache, and Mozilla are compatible with POCL.

**Using Pimcore Community Edition with Pimcore Studio:**
If you're combining the open-source Pimcore Community Edition (licensed under GPLv3) with Pimcore Studio (licensed under POCL), you need to separate functionalities.

For example:
- Server-side code (handling backend logic) can be GPLv3.
- Frontend code (related to Pimcore Studio) must stay under POCL.

The two parts should communicate via APIs and adhere to their respective licenses, ensuring the codebases remain distinct.

_Important GPLv3 Note:
Section 5 of GPLv3 (Verbatim Copying and Modified Versions) explains that any modified version of GPL software must also be distributed under the GPL. However, using an API to interact with GPL software is generally not considered modification or derivation, so APIs can help keep the two license types separate._

**Using Pimcore Enterprise Edition:**
If you're using the Pimcore Enterprise Edition, all components are licensed under the Pimcore Commercial License (PCL). This means you no longer have to worry about mixing different licenses like GPLv3 or POCL; everything falls under the PCL.

### GPLv3, POCL and PCL
Yes, you can distribute your extension based on POCL software (such as Pimcore Studio) without needing Pimcore's explicit acknowledgment, as long as you comply with the POCL terms. However, there are some important distinctions:

**1. POCL (Pimcore Open Core License)**
You can distribute your extension under the POCL by following these key conditions:
- Attribution and Compliance: You must properly attribute Pimcore and ensure that your extension complies with the POCL terms.
- Derivative Marking: Your extension must be clearly marked as a derivative and licensed under the POCL (Section 3.3).
- Distribution Terms: You must provide a copy of the POCL with your extension and follow all the terms related to modification and distribution.
- No Acknowledgment Needed: You do not need explicit acknowledgment from Pimcore to distribute an extension under the POCL, as long as you follow these rules.

However, if you intend to contribute back to the core POCL software (e.g., Pimcore Studio), you must adhere to the Pimcore Contributor License Agreement (CLA), as stated in Section 3.1.

**2. PCL (Pimcore Commercial License)**
When it comes to the Pimcore Commercial License (PCL), things are different:

- PCL is proprietary: It is a commercial license controlled by Pimcore GmbH, and distributing your extension under the PCL requires explicit approval or acknowledgment from Pimcore.
- No free licensing: You cannot independently license your extension under PCL, as it is a commercial license governing the Pimcore Enterprise Edition. Using PCL requires specific agreements with Pimcore, and any commercial distribution must be coordinated with us. Pimcore's upcoming enterprise Marketplace will feature PCL and PCL-compatible extensions. Feel free to reach out to Pimcore for any further questions.

> Licenses are a very weird and confusing topic.

Yes, Dominik, you’re completely right here. Licensing is a wild and complex topic, especially in the open-source world! It’s all about ownership, rights, and ultimately, intellectual property. Getting everything spot-on in terms of licensing has been our mission for the past few years. Everyone wants their assurance—both our commercial customers and the users of the Pimcore Community Edition.

And then there’s GPLv3, with its copyleft requirements, which adds another layer of complexity because of the need to convey changes and keep everything open. But all of those issues have been resolved months ago. Now, we’ve got it streamlined: PCL governs our enterprise offerings, POCL is in place for Pimcore Studio—because we’ve always wanted to keep Pimcore Studio free of charge—and GPLv3 and PCL together govern the open-source core of Pimcore.

That said, I’m not a legal advisor! So, for deeper, more nuanced questions, it’s always best to consult a professional. What I’ve outlined here is just a summary of what’s already defined in the licenses.

Cheers,
Dietz

**Reply by dpfaffenbauer** [reply]
thanks @drietsch. As always a perfect response :) That makes the situation more clear for everyone.

**Reply by dpfaffenbauer** [reply]
@drietsch one more question about licensing.

If I got a PCL and install a community extension which is GPL, is the whole project GPL again?

**Reply by dpfaffenbauer** [reply]
Point 4.1 of the PCL might imply that this is a violation of the PCL. Not sure honestly....

**Reply by drietsch** [reply]
> @drietsch one more question about licensing.
> 
> If I got a PCL and install a community extension which is GPL, is the whole project GPL again?

Yes, that’s correct. If you install a GPL-licensed community extension in a project governed by PCL, the GPL licensing terms will extend to the entire project. This is because the GPL is a copyleft license, meaning that any derivative work or combination with GPL-licensed code would need to be distributed under the same GPL terms. This could potentially mean you’d have to open-source the entire project, which would conflict with the commercial nature of the PCL. So, be cautious when mixing licenses like GPL with PCL.

**Reply by drietsch** [reply]
> Point 4.1 of the PCL might imply that this is a violation of the PCL. Not sure honestly....

4.1 The Customer shall not (a) use GPLv3-licensed Pimcore Community Edition alongside PCL licensed 
Software, and shall not (b) revert from PCL to GPLv3, to protect the Customer’s rights in 
Derivatives. 

It’s simply not legally feasible. The only solution is dual-licensing, as we've done with Pimcore by offering GPLv3, PCL, and POCL.

**Reply by dpfaffenbauer** [reply]
> The only solution is dual-licensing, as we've done with Pimcore by offering GPLv3, PCL, and POCL.

so every community extension could create its own license and is then compatible?

**Reply by drietsch** [reply]
> > The only solution is dual-licensing, as we've done with Pimcore by offering GPLv3, PCL, and POCL.
> 
> so every community extension could create its own license and is then compatible?

It’s not that straightforward in most cases. If your community extension has received open-source contributions under GPLv3 from others, you need to ensure that these contributors also agree to the "additional" license. This is why a Contributor License Agreement (CLA) is essential. Without it, dual-licensing isn't practically feasible.

**Reply by jdreesen** [reply]
So this basically means that when you use the enterprise edition, you're locked in and are not allowed to use (most of the) community bundles anymore? That's very unfortunate!

**Reply by dpfaffenbauer** [reply]
> It’s not that straightforward in most cases. If your community extension has received open-source contributions under GPLv3 from others, you need to ensure that these contributors also agree to the "additional" license. This is why a Contributor License Agreement (CLA) is essential. Without it, dual-licensing isn't practically feasible.

sure. Reason I am asking is: I am thinking about creating a Open License as a Community Service that is compatible with PCL. (Also requires a PCL)

That would solve the Community Extension Issue in PCL environments. Reason actually is that CoreShop requires a Community Extension, therefore CoreShop currently is PCL incompatible and either has to remove that dependency or find a solution that that extension is also dual licensed.

**Reply by drietsch** [reply]
> So this basically means that when you use the enterprise edition, you're locked in and are not allowed to use (most of the) community bundles anymore? That's very unfortunate!

No, that’s not actually the case. It depends on the license of the community bundle. If the bundle uses a license that is compatible with both GPLv3 and PCL—like MIT, Apache, or BSD—then it’s perfectly fine to use it with the enterprise edition. The issue arises when the bundle uses a copyleft license, such as GPLv3, because it requires that the source code be conveyed, potentially compromising your proprietary work.

So, the concern isn't about being "locked in" but rather whether the license is copyleft or copyright. Many companies become more cautious when they understand that using GPLv3 could put their copyright and intellectual property at risk, which is why it's such a significant consideration.

**Reply by drietsch** [reply]
Software licensing, especially on a global scale, is a complex task, and it's crucial to get it right from the start; otherwise, it can cause issues down the road. Here’s my personal take: I’m fully aware that most developers and coders don’t pay much attention to software licensing because it’s often seen as a mess—particularly when it involves intellectual property and liabilities. However, as a global company like Pimcore, which offers both the free Pimcore Community Edition and the Enterprise Edition, we have to be 120% certain that we’re doing it right.

Ultimately, it’s about protecting our customers and ensuring the people building solutions with Pimcore are safe from potential licensing risks. If they’re comfortable with GPLv3, that’s great. But we also need to provide the assurance that customers who can’t accept copyleft licenses like GPLv3 aren’t exposed to any risks. These are the two sides of the story. And Pimcore isn’t alone in this challenge—just look at WordPress, Joomla, Drupal, TYPO3, Akeneo (Community Edition), Magento (Community Edition), Shopware, WooCommerce, PrestaShop, and many more. They all face the same licensing dilemma.

**Reply by jdreesen** [reply]
> No, that’s not actually the case. It depends on the license of the community bundle. If the bundle uses a license that is compatible with both GPLv3 and PCL—like MIT, Apache, or BSD—then it’s perfectly fine to use it with the enterprise edition. The issue arises when the bundle uses a copyleft license, such as GPLv3, because it requires that the source code be conveyed, potentially compromising your proprietary work.

So it is possible to create a Pimcore bundle and *not* license it under GLPv3, even though Pimcore code is used directly in this bundle (without API in between)?
I assumed that the GPLv3 license of Pimcore also enforces a GPL license in bundles, since these are derived works?

**Reply by drietsch** [reply]
> > No, that’s not actually the case. It depends on the license of the community bundle. If the bundle uses a license that is compatible with both GPLv3 and PCL—like MIT, Apache, or BSD—then it’s perfectly fine to use it with the enterprise edition. The issue arises when the bundle uses a copyleft license, such as GPLv3, because it requires that the source code be conveyed, potentially compromising your proprietary work.
> 
> So it is possible to create a Pimcore bundle and _not_ license it under GLPv3, even though Pimcore code is used directly in this bundle (without API in between)? I assumed that the GPLv3 license of Pimcore also enforces a GPL license in bundles, since these are derived works?

Yes, it is possible to create a Pimcore bundle and not license it under GPLv3, even if Pimcore code is used directly. The key here is whether the Pimcore Community Edition (which is GPLv3) or the Enterprise Edition (which uses PCL) is involved.

If you're working with the Pimcore Community Edition under GPLv3, **any derived works**—including bundles—would generally need to comply with GPLv3, meaning they would also have to be licensed under GPLv3. This is because GPLv3 is a copyleft license, which enforces the same license on any derived works.

However, if you're developing for the Pimcore Enterprise Edition, which is licensed under PCL, you're not bound by GPLv3 restrictions, so you have more flexibility to choose other licenses, such as MIT, Apache, or BSD, as long as they are compatible with PCL. This is how dual licensing in Pimcore works: while the community version is GPLv3, the enterprise version provides a commercial licensing option that allows for more flexibility in terms of licensing derived works, like bundles.

**Reply by drietsch** [reply]
So, to elaborate: If you have an open-source Pimcore Community Edition bundle and want it to be fully compliant with all Pimcore Editions, you would need to dual-license it. This means using GPLv3 for the Pimcore Community Edition and a PCL-compliant license for Enterprise Edition.

**Reply by jdreesen** [reply]
So this means that each bundle must be dual-licensed (e.g., GPLv3 *and* MIT) if it should be usable with community *and* enterprise versions of Pimcore.

**Reply by drietsch** [reply]
Yes.

**Reply by jdreesen** [reply]
Maybe you should add this information to the Bundle Developer's Guide in the docs, because as you say, it's important to get it right from the start, because once it's released, someone might contribute to it, which would make it difficult to change the license later.

**Reply by dpfaffenbauer** [reply]
You also have to define (in your License File) that MIT is only used if the Pimcore is PCL.

**Reply by Gitman007** [reply]
@drietsch 
I'm still confused about this.
To be very specific: If I build an application on Pimcore, let's say "Invoicing for real estate agencies", can I use a single Pimcore installation under POCL to offer my invoicing application to multiple real estate agencies?

**Reply by dpfaffenbauer** [reply]
If it's considered a SASS, you can't. SASS I would say is a multi tenant application where you create multiple tenants or they can self-signup for it

**Reply by BerndDultinger** [reply]
> @drietsch I'm still confused about this. To be very specific: If I build an application on Pimcore, let's say "Invoicing for real estate agencies", can I use a single Pimcore installation under POCL to offer my invoicing application to multiple real estate agencies?

@Gitman007 - The Community Edition is free of charge for for everyone in Non-Production-Use. If you need Enterprise Edition features for the development, our partners can use the Pimcore Developer License.

If the Community Edition holds all capabilities needed and your client is below €5Mio in revenue, you can develop and deploy at no cost from Pimcore.
In case the client is above the €5Mio treshold then the Professional Edition is needed for Production Use.
In case the client needs Pimcore EE features, a license for EE is needed for Production Use.

And yes, if you are hosting this 1:1 for a client, you can do see, see above license limitation (revenue, features)

If you plan to build a SaaS service with it, you will need an OEM License from Pimcore, which all are based on the EE features.

**Reply by drietsch** [reply]
@Gitman007 I'd like to add to that SaaS topic. The POCL states:

> 2.2 The Customer may not offer the Open Core Software as a hosted or managed service by granting third parties access to a significant part of the features or functions of the Open Core Software.

It's important that the **"significant parts of the features or functions"** relate directly to Pimcore itself. This distinction clarifies the difference between offering Pimcore (or the majority of its functionalities) as a hosted service, and creating a SaaS application **based on** Pimcore (in your case, the Community Edition) for a specific industry use case — for example, an invoicing solution for real estate agencies.

To me, this sounds like a solution **built on top of Pimcore**, where users don’t actually interact with Pimcore or its admin interface directly.  Just my two cents.

Cheers,
Dietz

**Reply by Gitman007** [reply]
Thanks all for your responses!

But.. getting rather contradictory answers, even from within Pimcore, and "it sounds like" it's allowed, doesn't really reassure me.

Let me give an other concrete example: Under POCL, can I use a single Pimcore installation for rather simple static websites that I build for multiple clients of mine and charge them for the hosting? Can I let them login in Studio UI to edit some wysiwyg blocks? If yes, can they upload/change some pictures? Add a page? Translations... In other words, Where is the limit of "significant parts of the features or functions"? Everything they do is pure Pimcore functionality in this case.

Now I think of it, 2.2 Doesn't mention something like "on the same instance". So spinning up 50 Pimcore instances with exactly the same (Invoicing for real estate) application and charging for them 1:1 like BerndDultinger mentioned that is allowed vs 50 of those clients on the same Pimcore instance should not make any difference, not? (The more I think of it, the more I get confused maybe I should stop it for today :-) )

What also comes to mind: Let's say one of my examples is allowed under POCL, how would that €5Mio customer revenue threshold work when multiple clients are on the same Pimcore instance? The biggest client? My revenue? The sum of all customers revenue on that Pimcore instance? If most customers are small companies that don't have to publish their numbers, there is no way of knowing this.

Maybe let me also explain where I'm coming from.
I helped building a project, mostly frontend, when 2024.4 was the latest version. Before, I knew about Pimcore's existence but this was my first hands-on experience.
I was super enthousiast about it and was planing to use Pimcore for everything (and moving away from Wordpress). A couple of months later I discovered the new POCL and the Admin Classic to Studio UI switch. I don't want to get into the licensing discussion but let me just say that this dampened my enthusiasm considerably.

I was convinced that offering SaaS based on Pimcore under POCL wasn't allowed, a colleague was convinced otherwise, so I came back here.

And btw, I absolutely understand that Pimcore is overkill for a lot of what I do but I figured that also using it for small projects is a great way to learn and a path to grow into Pimcore to use it for bigger projects.
